### PR TITLE
[INTEGRATION] Shared config

### PR DESCRIPTION
**Pull request by TheSonOfThomp**
Creates the following packages
• `@lg-tools/build`: Configuration for building packages. Also includes `TSDoc` build script
• `@lg-tools/create`: Creates a new `@leafygreen-ui` package (improvements to come later)
• `@lg-tools/test`: Configuration for jest tests
• `@lg-tools/lint`: Configuration for esLint & Prettier
• `@lg-tools/storybook`: Configuration for Storybook
• `@lg-tools/storybook-decorators`: Decorators for Storybook
• `@lg-tools/slackbot`: Release announcements slackbot
• `@lg-tools/update`: Updates all `@leafygreen-ui` packages (improvements to come later)
• `@lg-tools/validate`: Build & Dependency validation scripts

Additionally, creates a single package (`@lg-tools/cli`) that builds a Command Line interface. 
This package is intended to be shared across other LeafyGreen repositories, as well as used in projects that use LeafyGreen UI components.

Current supported subcommands include: 

```
lg <subcommand>
  create [options] <name>                 Creates a new package with the provided name
  install [options] [packages...]         Installs LeafyGreen packages to the current app
  link [options] <destination>            Link local LeafyGreen packages to a destination app.
  unlink [options] <destination>          Unlink local LeafyGreen packages from a destination app.
  lint [options]                          Checks code and package.json formatting
  test [options]                          Tests leafygreen-ui packages with unified config.
  update|upgrade [options] [packages...]  Updates installed LeafyGreen packages based on the specified range
  validate [options]                      Validates build integrity, and package.json dependencies
  build-package                           Builds a package
  build-ts                                Builds a package's TypeScript definitions
  build-tsdoc                             Builds a package's TSDoc file
  help [command]                          display help for command
``` 

### Architecture

Inside `tools/cli` we use Commander to create the `lg` command, and all subcommands. 

`cli` depends on several other `@lg-tools` packages. When a subcommand is called, we fire an action function defined in one of the other `@lg-tools/` dependencies.

We define a `bin` property in `cli/package.json` which declares what script to run when we call `> lg` from the terminal

## How to test
Ensure each of the following commands runs without errors (most of these commands are run by CI)
- `yarn build`
- `yarn build:docs`
- `yarn start`
- `yarn lint`
- `yarn run link <path>`
- `yarn test` & `yarn test:ssr`
- `yarn run unlink <path>`
- `yarn validate`

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -12,93 +12,44 @@
     "build": "turbo run build tsc",
     "build-storybook": "storybook build",
     "build:docs": "npx ts-node ./scripts/parse-tsdocs.ts",
-    "lint": "npm-run-all --parallel lint:*",
-    "lint:prettier": "prettier --check \"**/*.{js,ts,tsx,json,md,yml}\"",
-    "lint:eslint": "eslint \"**/*.{js,ts,tsx}\"",
+    "lint": "lg-lint",
```
Comment (bruugey) [inline package.json:15]: When running this locally I get:

```
Checking formatting...
coverage/coverage-final.json
Oops! Something went wrong! :(

ESLint: 8.43.0

Error: Cannot read config file: /Users/brooke.yalof/mdb/leafygreen-ui/.eslintrc.js
Error: Cannot find module '@lg-tools/lint/eslint.config'
Require stack:
- /Users/brooke.yalof/mdb/leafygreen-ui/.eslintrc.js
- /Users/brooke.yalof/mdb/leafygreen-ui/node_modules/@eslint/eslintrc/dist/eslintrc.cjs
- /Users/brooke.yalof/mdb/leafygreen-ui/node_modules/eslint/lib/cli-engine/cli-engine.js
- /Users/brooke.yalof/mdb/leafygreen-ui/node_modules/eslint/lib/eslint/eslint.js
- /Users/brooke.yalof/mdb/leafygreen-ui/node_modules/eslint/lib/eslint/index.js
- /Users/brooke.yalof/mdb/leafygreen-ui/node_modules/eslint/lib/cli.js
- /Users/brooke.yalof/mdb/leafygreen-ui/node_modules/eslint/bin/eslint.js
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:1026:15)
    at Function.Module._load (node:internal/modules/cjs/loader:871:27)
    at Module.require (node:internal/modules/cjs/loader:1098:19)
    at require (node:internal/modules/cjs/helpers:108:18)
    at Object.<anonymous> (/Users/brooke.yalof/mdb/leafygreen-ui/.eslintrc.js:2:16)
    at Module._compile (node:internal/modules/cjs/loader:1196:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1250:10)
    at Module.load (node:internal/modules/cjs/loader:1074:32)
    at Function.Module._load (node:internal/modules/cjs/loader:909:12)
    at Module.require (node:internal/modules/cjs/loader:1098:19)
coverage/lcov-report/prettify.js
```

Comment (bruugey) [reply package.json:15]: (it's nested, just have to update a path somewhere I believe)

File: package.json [hunk 2]
```diff
@@ -12,93 +12,44 @@
...
     "chromatic": "npx chromatic",
     "clean": "yarn clean:builds && yarn clean:modules",
     "clean:builds": "rm -rf packages/*/dist packages/*/tsconfig.tsbuildinfo ./tsconfig.tsbuildinfo packages/**/stories.js",
     "clean:modules": "rm -rf node_modules packages/*/node_modules ",
-    "fix": "npm-run-all --parallel fix:*",
-    "fix:prettier": "prettier --write \"**/*.{js,ts,tsx,json,md,yml}\"",
-    "fix:eslint": "eslint \"**/*.{js,ts,tsx}\" --fix",
+    "fix": "lg-lint --fix",
```
Comment (bruugey) [inline package.json:20]: Seeing same as above (because it's calling `lint` first)

File: package.json [hunk 3]
```diff
@@ -12,93 +12,44 @@
...
     "publish": "npm-run-all --serial publish:*",
     "publish:packages": "yarn changeset publish --public",
-    "slackbot": "npx ts-node ./scripts/slackbot",
+    "slackbot": "slackbot",
     "start": "storybook dev -p 9001 --no-version-updates",
-    "test": "npx ts-node ./scripts/test.ts",
-    "test:local": "yarn test",
-    "test:ssr": "yarn test --ssr",
+    "test": "lg-test",
+    "test:ssr": "lg-test --ssr",
```
Comment (bruugey) [inline package.json:30]: this script not working either, but I have a feeling we'll be able to remove. I'm not entirely convinced this is providing enough value for the maintenance burden introduced.

```
☁  leafygreen-ui [integration-shared-config] yarn test:ssr
yarn run v1.22.19
$ lg-test --ssr
(node:79659) Warning: Accessing non-existent property 'setupFiles' of module exports inside circular dependency
(Use `node --trace-warnings ...` to show where the warning was created)
TypeError: defaultConfig.setupFiles is not iterable
    at Object.<anonymous> (/Users/brooke.yalof/mdb/leafygreen-ui/tools/test/config/ssr/jest.config.js:6:22)
    at Module._compile (node:internal/modules/cjs/loader:1196:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1250:10)
    at Module.load (node:internal/modules/cjs/loader:1074:32)
    at Function.Module._load (node:internal/modules/cjs/loader:909:12)
    at Module.require (node:internal/modules/cjs/loader:1098:19)
    at require (node:internal/modules/cjs/helpers:108:18)
    at readConfigFileAndSetRootDir (/Users/brooke.yalof/mdb/leafygreen-ui/node_modules/jest-config/build/readConfigFileAndSetRootDir.js:119:22)
    at readConfig (/Users/brooke.yalof/mdb/leafygreen-ui/node_modules/jest-config/build/index.js:210:65)
    at readConfigs (/Users/brooke.yalof/mdb/leafygreen-ui/node_modules/jest-config/build/index.js:406:32)
✨  Done in 0.28s.
```

File: tools/build/config/rollup.config.mjs [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+  output: {
+    dir: moduleFormatToDirectory[format],
+    name,
+    format,
+    sourcemap: true,
+    globals,
+    interop: 'compat', // https://rollupjs.org/configuration-options/#output-interop
+  },
+  plugins: [
+    nodeExternals({ deps: true }),
```
Comment (bruugey) [inline tools/build/config/rollup.config.mjs:132]: nit, but let's be consistent and line break in between each plugin

Comment (bruugey) [reply tools/build/config/rollup.config.mjs:132]: (or none of them)

File: tools/lint/config/npmpackagejsonlintrc.config.js [hunk 5]
```diff
@@ -0,0 +1,13 @@
+module.exports = {
+  rules: {
+    'prefer-caret-version-dependencies': 'warning',
+  },
+  overrides: [
+    {
+      patterns: ['./packages/code/package.json'],
```
Comment (bruugey) [inline tools/lint/config/npmpackagejsonlintrc.config.js:7]: is there a different way for us to override? this won't work in other packages, will it?

File: tools/lint/config/eslint.config.js [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+    },
+  },
+  env: {
+    es6: true,
+    browser: true,
+    node: true,
+  },
+  settings: {
+    react: {
+      version: '16.13.1',
```
Comment (bruugey) [inline tools/lint/config/eslint.config.js:26]: we'll have to update this sooon @shaneeza  :)

File: tools/lint/config/npmpackagejsonlintrc.config.js (hunk 5 above)
Comment (TheSonOfThomp) [reply tools/lint/config/npmpackagejsonlintrc.config.js]: 👍 Added specific dependency overrides

File: tools/build/config/rollup.config.mjs (hunk 4 above)
Comment (TheSonOfThomp) [reply tools/build/config/rollup.config.mjs:132]: Prettier looks at this file. IMO, if prettier doesn't complain about this, we should leave it

File: babel.config.js [hunk 7]
```diff
@@ -1,35 +1,2 @@
...
-    '@babel/plugin-proposal-logical-assignment-operators',
-    ['@babel/plugin-proposal-class-properties', { loose: true }],
-    ['@babel/plugin-proposal-private-methods', { loose: true }],
-    ['@babel/plugin-proposal-private-property-in-object', { loose: true }],
-    '@emotion/babel-plugin',
-  ];
-
-  return { presets, plugins };
-};
+const babel = require('@lg-tools/build/config/babel.config');
```
Comment (bruugey) [inline babel.config.js:1]: super nit but should we rename this `babelConfig`

File: package.json [hunk 8]
```diff
@@ -8,98 +10,48 @@
...
     "clean:builds": "rm -rf packages/*/dist packages/*/tsconfig.tsbuildinfo ./tsconfig.tsbuildinfo packages/**/stories.js",
     "clean:modules": "rm -rf node_modules packages/*/node_modules ",
-    "fix": "npm-run-all --parallel fix:*",
-    "fix:prettier": "prettier --write \"**/*.{js,ts,tsx,json,md,yml}\"",
-    "fix:eslint": "eslint \"**/*.{js,ts,tsx}\" --fix",
-    "link": "./scripts/link.sh",
-    "link:all": "lerna exec -- yarn link",
+    "fix": "lg-lint --fix",
+    "link": "lg-link",
+    "lint": "lg-lint",
```
Comment (bruugey) [inline package.json:23]: do we want a lint script that doesn't fix?

File: package.json [hunk 9]
```diff
@@ -8,98 +10,48 @@
...
     "prepublishOnly": "yarn fix && yarn test && yarn build && yarn prepublish:ts-downlevel",
     "prepublish:ts-downlevel": "rm -rf packages/*/dist/ts3.4 && echo packages/*/dist | NODE_NO_WARNINGS=1 xargs -I{} -n1 -P8 downlevel-dts {} {}/ts3.4",
     "publish": "npm-run-all --serial publish:*",
     "publish:packages": "yarn changeset publish --public",
```
Comment (bruugey) [inline package.json:27]: can we combine the two publish scripts into one?

File: tools/create/package.json [hunk 10]
```diff
@@ -0,0 +1,20 @@
+{
+  "name": "@lg-tools/create",
```
Comment (bruugey) [inline tools/create/package.json:2]: i feel like if we're introducing this we should at least explore using turbo gen

File: package.json (hunk 8 above)
Comment (TheSonOfThomp) [reply package.json:23]: Useful for CI

File: package.json (hunk 9 above)
Comment (TheSonOfThomp) [reply package.json:27]: Maybe... I haven't dug into them, so I'm not 100% sure what they're doing

File: package.json [hunk 11]
```diff
@@ -8,98 +10,48 @@
...
-    "slackbot": "npx ts-node ./scripts/slackbot",
+    "slackbot": "slackbot",
```
[the remark this answers is not in the document]
Comment (TheSonOfThomp) [reply package.json:28]: Not necessarily, we can call it whatever we want. I'll change it to lg-slackbot though for clarity 👍

File: .github/workflows/pr.yml [hunk 12]
```diff
@@ -62,6 +41,37 @@ jobs:
         if: ${{ steps.build-cache.outputs.cache-hit != 'true' }}
         run: yarn build
 
+  lint:
```
Comment (bruugey) [inline .github/workflows/pr.yml:44]: have you noticed an improvement in CI job time?

**Review by bruugey — COMMENTED** [review]
One quick nit.. I feel like we're moving towards a world of named exports only. How would you feel about making all of the exports here named before merging? if you feel strongly, don't me block you - I do think this is the way we're moving though.

File: tools/create/package.json [hunk 13]
```diff
@@ -11,7 +11,8 @@
     "build": "rollup -c rollup.config.mjs"
   },
   "publishConfig": {
-    "access": "public"
+    "access": "public",
```
Comment (bruugey) [inline tools/create/package.json:14]: rm public while its artifactory?

**Review by bruugey — COMMENTED** [review]
I think we should add a changeset for all of the changed packages so if something breaks we can roll it back more easily (or at least properly handle versioning)

**Comment by TheSonOfThomp** [comment]
> One quick nit.. I feel like we're moving towards a world of named exports only. How would you feel about making all of the exports here named before merging? if you feel strongly, don't me block you - I do think this is the way we're moving though.

As we discussed at sync, these default exports are required by the build tools (`rollup`, `babel` etc).

**Comment by TheSonOfThomp** [comment]
> Also - did we decide to handle the rollup config differently (a default that's overrid-eable rather than a config in every package)

@bruugey This work is included in #1846. Thoughts on merging these bodies of work?

**Review by bruugey — APPROVED** [review]

**Pull request merged**